### ISSUE TITLE
Allow users to opt out of validation messages and emails

### ISSUE THREAD

**Issue by bgirardot** [hotosm/tasking-manager#635]
Experienced users might wish to no longer get validation / invalidation emails.

Based on feedback I changed this to be opting out of the automated validation tasking manager messages totally as well as the email notices about those messages.

**Comment by OttoROSM** [comment]
Hi Blake,
Came here to report this. At least do not send a standardized message to the e-mail address that there is a message in the tasking manager. Especially when the message in the Tasking Manager is a standard message where someone validated a tile.

Typically I would map multiple tiles after each other, so when someone else validates 10 in a row, I get 10 e-mails saying there is a message in the tasking manager for me.
Then I go to the tasking manager and need to open 10 times a similar message saying that my tile was validated and I need to keep up the good work. Then click delete for each message and confirm I want to delete each message and then go back to my e-mail inbox and delete 10 e-mails.

Rather I would receive the message directly in my inbox, or receive a summary report saying, 10 tiles were validated, or you have 10 messages in the tasking manager. 

Basically unless someone made a specific comment, I am fine with getting a simpel statistic saying 10/10 tiles you mapped were validated.

**Comment by verdy-p** [comment]
10 mails in the TM manager itself that need to be clicked separately and deleted by another click, it's too much. And yes I'm not necessarily interested to have such inbox mails, or even notifications.
- there should be options to disable these notifications or mails in the TM messsaging box
- also an option to not forward these with more notifications or mails to our regular mailbox for such events (we need filters by notification type to enable/disable them at their respective source)
- also a way to select multiple messages in the TM box to archive/delete/mark them as read without having to click each one.

**Comment by ethan-nelson** [comment]
For someone interested in working on this, here's a possible route that should cover most of the work:

1) In the users schema, add a database column for email notifications (https://github.com/hotosm/tasking-manager/blob/develop/server/models/postgis/user.py)
2) Add a corresponding field in the user data object for this column (https://github.com/hotosm/tasking-manager/blob/develop/server/models/dtos/user_dto.py)
3) Add a corresponding field in the user update api (https://github.com/hotosm/tasking-manager/blob/develop/server/api/users/user_apis.py#L52-L116)
4) Add a selectable option on the user profile to disable notifications (https://github.com/hotosm/tasking-manager/blob/develop/client/app/profile/profile.html)
5) In the messaging API, add a check to see if the subscribed column is set to false (https://github.com/hotosm/tasking-manager/blob/develop/server/services/messaging/message_service.py#L63, https://github.com/hotosm/tasking-manager/blob/develop/server/services/messaging/message_service.py#L108, https://github.com/hotosm/tasking-manager/blob/develop/server/services/messaging/message_service.py#L135)

**Comment by fiennyangeln** [comment]
hi I want to take this issue

**Comment by bgirardot** [comment]
@fiennyangeln People will be very happy if you create this feature :)

Please note: if you make a change to the database to allow storage of the opt out, you will need to generate a alembic script to migrate the database, please only do this on your local testing db during development :)

**Comment by bgirardot** [comment]
This might need some more thinking through as well.

Right now you get sent a message when your task is validated. If you have an email address, an email gets sent as well.

I think what people want is to not get any message, so if they do not get a message, the internal tasking manager message, then they will automatically not get an email either.

So I would focus on "opt out of validation messages" that way they get no message, no email, nothing when a task is validated.

They still need to get messages someone directed at them with @username , just not the automated validation messages.

**Comment by fiennyangeln** [comment]
@bgirardot please review

**Comment by bgirardot** [comment]
Please note, I think this issue has evolved, in addition to the comments above, others have complained about the TM messages for validation in general.

I think the best fix is to opt out of any validation messages then the emails would never be sent.

I am not sure, maybe we need both options?

1. Opt out of any auto generated validation messages
2. Opt out of emails notices of validation messages

**Comment by PaulUithol** [comment]
@bgirardot see this is already an issue, sorry for email spamming you about this separately ;).

From my perspective (having gotten 80 or so validation emails over the past week), this is not a very helpful feature at the moment. Primarily because of the sheer volume if you're mapping slightly fanatically for a while, secondarily because the emails give you no hint or clue whatsoever what they're about.

I would like to get messages containing a) useful feedback on specific tiles, b) invalidated tiles, or c) answers to questions/mentions for tasks as a whole. But not for any square that's basically fine... But even then, I'd definitely want have that content mostly embedded directly in the email, including direct link to the relevant task/square?

**Comment by JesseWeinstein** [comment]
I heard from @bgirardot that the development team wants to leave the feature of sending emails about validated tasks on by default, but is open to adding an option for people to opt-out. Is this already being worked on, or would it be useful for me to create a PR to implement that?

**Comment by bgirardot** [comment]
@JesseWeinstein This has been completed and is in the production code now deployed on the HOT Tasking Manager. thank you for your offer! There is more that needs doing in this repo if you are still interested. please email me blake.girardot@hotosm.org or join our slack http://slack.hotosm.org/
